bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Bookie.java

```diff
@@ -441,117 +335,187 @@ private void checkIfDirsOnSameDiskPartition(List<File> dirs) throws DiskPartitio
-                    if (journalCookie.isBookieHostCreatedFromIp()) {
-                        conf.setUseHostNameAsBookieID(false);
-                    } else {
-                        conf.setUseHostNameAsBookieID(true);
-                    }
-                } catch (FileNotFoundException fnf) {
-                    newEnv = true;
-                    missedCookieDirs.add(journalDirectory);
-                }
-            }
-
+            // 1. retrieve the instance id
             String instanceId = rm.getClusterInstanceId();
+
+            // 2. build the master cookie from the configuration
             Cookie.Builder builder = Cookie.generateCookie(conf);
             if (null != instanceId) {
                 builder.setInstanceId(instanceId);
             }
             Cookie masterCookie = builder.build();
-            Versioned<Cookie> rmCookie = null;
-            try {
-                rmCookie = Cookie.readFromRegistrationManager(rm, conf);
-                // If allowStorageExpansion option is set, we should
-                // make sure that the new set of ledger/index dirs
-                // is a super set of the old; else, we fail the cookie check
-                masterCookie.verifyIsSuperSet(rmCookie.getValue());
-            } catch (CookieNotFoundException e) {
-                // can occur in cases:
-                // 1) new environment or
-                // 2) done only metadata format and started bookie server.
-            }
-            for (File journalDirectory : journalDirectories) {
-                checkDirectoryStructure(journalDirectory);
-            }
-            if (!newEnv) {
-                for (Cookie journalCookie: journalCookies) {
-                    masterCookie.verifyIsSuperSet(journalCookie);
-                }
-            }
 
+            // 3. read the cookie from registration manager. it is the `source-of-truth` of a given bookie.
+            //    if it doesn't exist in registration manager, this bookie is a new bookie, otherwise it is
+            //    an old bookie.
+            final Versioned<Cookie> rmCookie = readAndVerifyCookieFromRegistrationManager(
+                masterCookie,
+                conf,
+                rm
+            );
+
+            // 4. check if the cookie appear in all the directories.
+            List<File> missedCookieDirs = new ArrayList<>();
+            List<Cookie> existingCookies = Lists.newArrayList();
+            if (null != rmCookie) {
+                existingCookies .add(rmCookie.getValue());
+            }
 
-            for (File dir : allLedgerDirs) {
-                checkDirectoryStructure(dir);
-                try {
-                    Cookie c = Cookie.readFromDirectory(dir);
-                    masterCookie.verifyIsSuperSet(c);
-                } catch (FileNotFoundException fnf) {
-                    missedCookieDirs.add(dir);
-                }
+            // 4.1 verify the cookies in journal directories
```

Comment:
How is this code ensuring that all directories from rmCookie are being checked and that the cookies are valid?
As I understand it, this is a byproduct of the fact that we called verifyIsSuperSet when reading rmCookie, but it takes a lot of digging to figure this out, and it's action from afar.

Also, do we allow journal expansion, or just ledger dir expansion?

Reply:
- masterCookie should be the super set of rmCookie. masterCookies is generated from configuration, so all the directories are covered.
- we don't allow journal expansion. the storage expansion was added before multiple journals.

Reply:
in the expansion case, masterCookie is a superset or rmCookie, so you'll end up with some directories missing cookies. It's not clear if these directories are part of rmCookie (i.e. preexisting directories) or newly added directories.

> we don't allow journal expansion. the storage expansion was added before multiple journals.
This should be clear from reading the checkEnv method. Currently it is not

Reply:
```
It's not clear if these directories are part of rmCookie (i.e. preexisting directories) or newly added directories.
```
this is part of `verifyDirsForOldEnvironment`

```
This should be clear from reading the checkEnv method. Currently it is not
```

why do you need to know this from reading checkEnv. it should be part of the logic when handling old environment, which is 'verifyDirsForOldEnvironment'. All the features about 'expansion" should only be related to old environment.

Reply:
verifyDirsForOldEnvironment does take rmCookie directly (only indirectly from it having been added to existing cookie). So its unclear which cookie was the canonical cookie for the old environment (IMO it should be rmCookie).

Reply:
masterCookie is the canonical cookie. we are using masterCookie for all the validations. rmCookie is the source-of-truth of whether it is a new environment or an old environment. so that's why the sequence is:

- generating masterCookie from configuration
- get rmCookie so we know whether it is a new environment or old environment
- going through all the directories listed in masterCookie to collect existing cookies and directories that missed cookies.
- if no directory is missing cookie, this is a healthy environment. we return. otherwise, we are getting into the fixing logic. The fixing logic can be creating cookies for new environment and expanding storage directory for old environment.

The logic flow is very clear and it is only one main logic and no branches. Again, personally I don't like branches here. when I was first initially working on this change, I spent almost half an hour to understand what are the differences between two branches. all the code was just duplicated.

Reply:
ugh, I just realized that it's possible to have multiple cookies in the registration manager, and they can possibly be all different :/

Reply:
verifyDirsForOldEnvironment is a bad name, it should really be verifyDirsForStorageExpansion.

I think what's making me most uncomfortable about the algorithm is the fact that the check to storage expansion is hidden in the subcalls.

How about?
```
            // 5. if there is no directory missing cookie, nothing need to be changed.                                                                                                                                                                                                    
            if (!missedCookieDirs.isEmpty()) {
                if (null == rmCookie) {
                    fixupCookieForNewEnvironment(missedCookieDirs);
                } else if (conf.isStorageExpansionEnabled()) {
                    fixupCookieForStorageExpansion(missedCookieDirs, existingCookies);
                }

                // verify that all directories and zk contain a cookie matching the master cookie
           }
```

Reply:
if a user was switching between ip, hostname and advertised address, you can possibly have multiple cookies, because they are using different bookie ids. but this is fixed by my change. now it won't happen. 

if you are always using same ip address, it is not possible to have multiple cookies.

Reply:
1) the difference between fixupCookieForNewEnvironment and fixupCookieForStorageExpansion are same at the actual fixing steps. They have to do the exactly same thing (even we add journal expansion in future). The only difference is on verification. I am fine if you want to combine verification and the actual fixing step together as a single 'fixup' call.

2) I am not sure why do you need to verify all the directories again with master cookie. all the validation should already happen before fixup. if fixup failed, exceptions should already be thrown. I am fine with doing another around of verification by recursively calling the checkEnvironment again, there should be no duplicated verification code again.